Fix #1020: Rendering of Svg images from server.

veena14cs opened the pull request:
## Explanation
This PR  fixes issue #1020. It implements rendering of svg images from server.

For Testing
1) In `GlideImageLoader` on line number 49 replace `model` with url "https://cdn.shopify.com/s/files/1/0496/1029/files/Freesample.svg"
2) In assets folder goto `fraction_exploration0.json` file. Search for image name `img_20180121_113315_pqwqhf863w_height_565_width_343` and replace png to svg.

veena14cs committed:
Added Svgencoder

veena14cs committed:
Update UrlImageParser.kt

rt4914 commented on utility/src/main/java/org/oppia/util/parser/GlideImageLoader.kt:
```diff
@@ -2,18 +2,22 @@ package org.oppia.util.parser
-import javax.inject.Inject
 import org.oppia.util.caching.AssetRepository
 import org.oppia.util.caching.CacheAssetsLocally
+import javax.inject.Inject
```
This should come above otherwise it will give lint issues.

veena14cs replied:
Done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/RepositoryGlideModule.kt:
```diff
@@ -1,17 +1,25 @@
-import java.io.InputStream
+import com.caverock.androidsvg.SVG
 import org.oppia.util.caching.AssetRepository
+import java.io.InputStream
```
organise imports for lint checks

veena14cs replied:
Done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/RepositoryGlideModule.kt:
```diff
@@ -1,17 +1,25 @@
 @GlideModule
 class RepositoryGlideModule : AppGlideModule() {
+//  override fun registerComponents(context: Context, glide: Glide, registry: Registry) {
+//    registry.prepend(ImageAssetFetcher::class.java, InputStream::class.java, RepositoryModelLoader.Factory())
```
Remove this code if not needed

veena14cs replied:
I want @BenHenning approval on this. As I have made changes in RepositoryGlideModule. Will this affect any other implementation?

rt4914 replied:
I think it is affecting other modules as your non-flaky tests are failing.

veena14cs replied:
I will check this then.

BenHenning replied:
Don't we still need this in order to properly support offline bitmap assets? We should retain resource loading for binary images, and also add that support for SVG.

veena14cs replied:
I need to add support for both. But I am not getting clear idea on how I can add svg support for GlideModule along with ImageAssetFetcher.

veena14cs replied:
@BenHenning I tried using append function to support offline loading of images like 
      `append(ImageAssetFetcher::class.java, InputStream::class.java, RepositoryModelLoader.Factory())`. and it is working.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/SvgDecoder.kt:
```diff
@@ -0,0 +1,31 @@
+import java.io.IOException
+import java.io.InputStream
+
+class SvgDecoder : ResourceDecoder<InputStream?, SVG?> {
```
Add kdoc for this class

rt4914 commented on utility/src/main/java/org/oppia/util/parser/SvgDrawableTranscoder.kt:
```diff
@@ -0,0 +1,20 @@
+import com.bumptech.glide.load.resource.transcode.ResourceTranscoder
+import com.caverock.androidsvg.SVG
+
+class SvgDrawableTranscoder : ResourceTranscoder<SVG?, PictureDrawable?> {
```
Ditto

rt4914 commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,28 @@
+import java.io.File
+import java.io.FileOutputStream
+
+class SvgEncoder : ResourceEncoder<SVG?> {
```
Ditto

veena14cs replied:
Done.

rt4914 commented on utility/src/main/java/org/oppia/util/parser/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,37 @@
+ * Listener which updates the [ImageView] to be software rendered, because [ ]/[Picture][android.graphics.Picture] can't render on a
+ * hardware backed [Canvas][android.graphics.Canvas].
+ */
+class SvgSoftwareLayerSetter : RequestListener<PictureDrawable> {
```
Remove commented code in this file if not needed.

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
Add test cases to add various public functions in this PR.

Otherwise as per your suggested changes your PR is working correctly. Thanks.

veena14cs commented:
> Add test cases to add various public functions in this PR.

There is issue #277 that need to have testcase for image loading. So I haven't added it here. Issue is mentioned in UrlImageParser file.

veena14cs committed:
updated kdoc

veena14cs committed:
Update SvgSoftwareLayerSetter.kt

aggarwalpulkit596 commented:
Are we handling the fallback case if the image is not an SVG?

veena14cs commented:
> Are we handling the fallback case if the image is not an SVG?

Yes.

aggarwalpulkit596 commented:
We can add one more case though to render the url as png if svg fails because some websites have a redirect mechanism in svg url which redirects to png url

aggarwalpulkit596 commented:
@veena14cs there is an error function in glide which executes if the image fails to render we can add png/svg clause in case of a failure

veena14cs commented:
> @veena14cs there is an error function in glide which executes if the image fails to render we can add png/svg clause in case of a failure

In that case you can create a PR and will check it.

veena14cs committed:
Update SvgEncoder.kt

rt4914 reviewed (changes requested):
Non-flaky tests are failing, else looks good.

nikitamarysolomanpvt reviewed (changes requested):
non_flaky_tests  are failing PTAL

veena14cs committed:
Merge branch 'develop' into render-svg

veena14cs committed:
Update HtmlParserTest.kt

veena14cs committed:
Update RepositoryGlideModule.kt

veena14cs committed:
Update RepositoryGlideModule.kt

veena14cs committed:
Update RepositoryGlideModule.kt

veena14cs commented:
> Non-flaky tests are failing, else looks good.

> non_flaky_tests are failing PTAL

PTAL its resolved now.

veena14cs committed:
Update SvgSoftwareLayerSetter.kt

veena14cs committed:
Update RepositoryGlideModule.kt

rt4914 reviewed (approved):
LGTM, thanks.

nikitamarysolomanpvt reviewed (approved):
LGTM

BenHenning commented on utility/src/main/java/org/oppia/util/parser/RepositoryGlideModule.kt:
```diff
@@ -1,17 +1,22 @@
   override fun registerComponents(context: Context, glide: Glide, registry: Registry) {
-    registry.prepend(ImageAssetFetcher::class.java, InputStream::class.java, RepositoryModelLoader.Factory())
+    registry.register(SVG::class.java, PictureDrawable::class.java, SvgDrawableTranscoder())
+      .prepend(SVG::class.java, SvgEncoder())
```
Do we actually need this encoder? Does Glide need it for caching the image?

veena14cs replied:
Yes we need it to write image to the stream. Otherwise image is not loaded.

BenHenning replied:
Ack thanks--I guessed it was for caching but I just wanted to be sure.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -60,7 +69,40 @@ class UrlImageParser private constructor(
+    }
+  }
+
+  private inner class SvgBitmapTarget(private val urlDrawable: UrlDrawable) :
```
Nit for formatting:

```kotlin
private inner class SvgBitmapTarget(
  private val urlDrawable: UrlDrawable
): CustomTarget<PictureDrawable>() {
  ...
}
```

I think that wraps a bit better, but I'm not sure how ktlint will handle this case.

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -60,7 +69,40 @@ class UrlImageParser private constructor(
+      resource: PictureDrawable,
+      transition: Transition<in PictureDrawable>?
+    ) {
+      val drawable = PictureDrawable(resource.picture)
```
Why create a copy of the ``PictureDrawable`` vs using the one being passed in? Is it because we're changing the drawable? If so, I suggest adding a brief comment explaining that the duplication is intentional & why since at a glance this could look like an accident that, when changed, might introduce bugs.

veena14cs replied:
No we can directly use drawable being passed. It was accidentally written.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -60,7 +69,40 @@ class UrlImageParser private constructor(
+    }
+  }
+
+  private inner class SvgBitmapTarget(private val urlDrawable: UrlDrawable) :
```
I think we can simplify the new class and consolidate with the old one. Consider:

```kotlin
  private inner class BitmapTarget(urlDrawable: UrlDrawable): CustomImageTarget<Bitmap>(
    urlDrawable, resource -> BitmapDrawable(context.resources, resource)
  )

  private inner class SvgTarget(urlDrawable: UrlDrawable): CustomImageTarget<PictureDrawable>(
    urlDrawable, resource -> PictureDrawable(context.resources, resource.picture)
  )

  private inner class CustomImageTarget<T>(
    private val urlDrawable: UrlDrawable, private val drawableFactory: (T) -> Drawable
  ) : CustomTarget<T>() {
    override fun onLoadCleared(placeholder: Drawable?) {
      // No resources to clear.
    }

    override fun onResourceReady(resource: T, transition: Transition<in T>?) {
      val drawable = drawableFactory(resource)
      htmlContentTextView.post {
        htmlContentTextView.width {
          val drawableHeight = drawable.intrinsicHeight
          val drawableWidth = drawable.intrinsicWidth
          val initialDrawableMargin = if (imageCenterAlign) {
            calculateInitialMargin(it, drawableWidth)
          } else {
            0
          }
          val rect = Rect(initialDrawableMargin, 0, drawableWidth + initialDrawableMargin, drawableHeight)
          drawable.bounds = rect
          urlDrawable.bounds = rect
          urlDrawable.drawable = drawable
          htmlContentTextView.text = htmlContentTextView.text
          htmlContentTextView.invalidate()
        }
      }
    }
  }
```

veena14cs replied:
Simplified this. Thanks.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,33 @@
+import com.bumptech.glide.request.target.Target
+
+/**
+ * Listener which updates the {@link ImageView} to be software rendered, because {@link
```
How does it do that? It seems from the implementation that the resource will never be ready, so it's not clear to me how it prevents hardware rendering.

veena14cs replied:
SVG format can not be hard decoding, it is soft decoding ImageView set at onResourceReady. I am also not so much aware of this. Need to dig in deep to understand.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * Listener which updates the {@link ImageView} to be software rendered, because {@link
+ * com.caverock.androidsvg.SVG SVG}/{@link android.graphics.Picture Picture} can't render on a
```
Is this a limitation in the SVG library? From the Android documentation ``Picture``s can be rendered on hardware-backed surfaces on SDK 23+. It seems worthwhile to allow that optimization for SDKs that support it. Or, did misunderstand the documentation?

https://developer.android.com/reference/android/graphics/Picture

veena14cs replied:
According to the PictureDrawable docs, Canvas doesn't support rendering Pictures/SVGs on hardware canvases until Marshmallow. I think it will be blurry in lower versions or may be there  will be performance issue. I need to test it.

BenHenning replied:
Can we mention in the docs that it's fine to use a hardware canvas for SDK versions 23+? That's more of what I was getting at--if it's actually possible, we should call out that it's allowed since we may want to do this in the future (e.g. check if we're running on SDK 23+ and, if so, use a hardware-accelerated layout to improve image rendering performance).

veena14cs replied:
Added in doc.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgDrawableTranscoder.kt:
```diff
@@ -0,0 +1,21 @@
+  override fun transcode(
+    toTranscode: Resource<SVG?>,
+    options: Options
+  ): Resource<PictureDrawable?>? {
```
What if we encoded from SVG to Picture, instead? Or, do we have to encode to a ``Drawable``? I don't see explicit value in using ``Drawable`` here since we just copy it later, and we don't seem to use ``Drawable``s for bitmaps until the final step (rendering it).

veena14cs replied:
Glide is using PictureDrawable class, so we need to pass drawable.

veena14cs replied:
Update from PictureDrawable to Picture.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgDecoder.kt:
```diff
@@ -0,0 +1,32 @@
+    options: Options
+  ): Resource<SVG?>? {
+    return try {
+      val svg = SVG.getFromInputStream(source)
```
Does/should this close the input stream? Glide's documentation hopefully indicates whether we're responsible for closing it.

veena14cs replied:
Closed input stream.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,29 @@
+    return try {
+      val svg: SVG = data.get()
+      val picture = svg.renderToPicture()
+      picture.writeToStream(FileOutputStream(file))
```
Similarly, do we need to close this output stream or does Android?

veena14cs replied:
Yes we need to close it. done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,29 @@
+    return try {
+      val svg: SVG = data.get()
+      val picture = svg.renderToPicture()
+      picture.writeToStream(FileOutputStream(file))
```
Where does ``writeToStream`` exist? I don't see it in Android's documentation for ``Picture``.

veena14cs replied:
It is Picture's writeToStream method it's undocumented. May be this SO answer can help us https://stackoverflow.com/a/5531302/4500099

BenHenning replied:
Ack, thanks. Given it's a private API, we ought to avoid using it since there may be no guarantee for compatibility. Once concern I'd have is that migrating data from one phone to another (e.g. after purchasing a new device) might cause compatibility issues here. Could we instead store the raw SVG file and load that each time?

veena14cs replied:
I found that it is deprecated [reference](https://emanual.github.io/Android-docs/reference/android/graphics/Picture.html) . And Document recommends to draw the picture into a Bitmap instead.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,29 @@
+      picture.writeToStream(FileOutputStream(file))
+      true
+    } catch (e: Exception) {
+      e.printStackTrace()
```
We should use our Logger class to log this rather than default logging it to ``System.err``.

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,29 @@
+  override fun encode(data: Resource<SVG?>, file: File, options: Options): Boolean {
+    return try {
+      val svg: SVG = data.get()
+      val picture = svg.renderToPicture()
```
Shouldn't we encode the actual SVG on disk rather than a bitmap representation of it? We want to retain the benefits of the image being scalable when we render it natively.

veena14cs replied:
It was easy to save a Picture object to disk so we copy contents of Svg to Picture..

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgDrawableTranscoder.kt:
```diff
@@ -0,0 +1,21 @@
+    options: Options
+  ): Resource<PictureDrawable?>? {
+    val svg: SVG = toTranscode.get()
+    val picture = svg.renderToPicture()
```
By pre-rendering it to a Picture does the SVG still scale without artifacts when rendered in a view?

veena14cs replied:
Scaling looks  fine. I tried loading high and low resolution images. It renders smoothly without blurring image.

BenHenning replied:
Hmm I guess because a picture object stores draw calls that this scales as expected. In that case, maybe it's best for us to pass a Picture along the pipeline since it probably will be a bit more performant to only render the picture once per load of the image rather than each time we actually want to render it.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/RepositoryGlideModule.kt:
```diff
@@ -1,17 +1,22 @@
+    registry.register(SVG::class.java, PictureDrawable::class.java, SvgDrawableTranscoder())
+      .prepend(SVG::class.java, SvgEncoder())
+      .append(InputStream::class.java, SVG::class.java, SvgDecoder())
+      .append(ImageAssetFetcher::class.java, InputStream::class.java, RepositoryModelLoader.Factory())
```
I think as a whole what you put together seems reasonable. I think we can maybe simplify it a bit further:
- We could create a custom type to represent an OppiaImage (this is a good time to make use of Kotlin's sealed classes--we want exactly one subtype of OppiaImage that corresponds to bitmaps, and a second for SVGs)
- We would then need to update ImageAssetFetcher, ImageLoader, and GlideImageLoader to reference OppiaImage instead of Bitmap
- We would probably need to introduce a mechanism in OppiaImage to convert itself from/to a ByteArray to work correctly with the asset system
- Some of the changes you made below in UrlImageParser and maybe here will likely need to change if we pass around an OppiaImage instead of a Bitmap

The above changes help make SVGs more of a 'first class citizen" rather than being sort of tacked onto the bitmap loading pipeline. That may help provide some simplifications in the future, as well.

veena14cs replied:
Do we need to introduce OppiaImage in this PR itself.

BenHenning replied:
If you can solve all of the other comments without making these changes, I think they're fine as a follow-up work item. If you choose to go that route, please file an issue to track it, add a TODO in this file for that issue, and please make sure all of the details above are captured in the issue.

veena14cs replied:
> Thanks @veena14cs. I followed up to the comments I saw that had open questions--please let me know if I missed any or if you need additional input on anything.

@BenHenning I have addressed all of your comments. Please take a glance once and let me know if any further changes are required.

veena14cs replied:
I have created issue #1039  to track this and added todo comment in the code.

BenHenning replied:
Thanks. Please format the TODO as ``TODO(#1039): Introduce custom type...``

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -36,11 +37,19 @@ class UrlImageParser private constructor(
-      gcsPrefix + gcsResource + imageUrl,
-      target
-    )
+    if (imageUrl.endsWith("svg")) {
```
I suggested above using a common OppiaImage so that you need just one load method in the image loader. Can't Glide identify an SVG image or do we actually need to do it?

veena14cs replied:
As we have separate implementation for loading bitmap and svg I have put a check over here.

BenHenning reviewed (commented):
Thanks @veena14cs! Overall I think this looks really good. I had some ideas to simplify it further--let me know if you agree with these, and send this back my way when it's ready for a follow-up review.

veena14cs committed:
Update UrlImageParser.kt

veena14cs commented:
@BenHenning I took this sample Svg code from here https://github.com/bumptech/glide/tree/master/samples/svg/src/main/java/com/bumptech/glide/samples/svg and https://stackoverflow.com/a/51206282/4500099.

I don't have enough knowledge on this, so I have addressed some of your comments please check and correct me wherever needed.

veena14cs committed:
update changes.

veena14cs committed:
Merge branch 'develop' into render-svg

BenHenning commented:
Thanks @veena14cs. I followed up to the comments I saw that had open questions--please let me know if I missed any or if you need additional input on anything.

veena14cs committed:
updated changes.

veena14cs commented:
> Thanks @veena14cs. I followed up to the comments I saw that had open questions--please let me know if I missed any or if you need additional input on anything.

@BenHenning I have addressed all of your comments. Please take a glance once and let me know if any further changes are required.

veena14cs committed:
updated nit changes

BenHenning commented:
Thanks @veena14cs--in the future please assign me the pull request once you need some sort of action item from me (e.g. answering a question or doing another review) so that I can find it more easily. Thanks! :)

I will try to take a look today.

veena14cs commented:
> Thanks @veena14cs--in the future please assign me the pull request once you need some sort of action item from me (e.g. answering a question or doing another review) so that I can find it more easily. Thanks! :)
> 
> I will try to take a look today.

Sure @BenHenning  thanks.

veena14cs committed:
Fixed lint errors

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgEncoder.kt:
```diff
@@ -0,0 +1,43 @@
+      val os = ByteArrayOutputStream()
+      bitmap.compress(Bitmap.CompressFormat.PNG, /* quality= */100, os)
+      bitmap.recycle()
+      os.writeTo(FileOutputStream(file))
```
I don't think that we want this, either. We should be storing the raw SVG file that we downloaded, not a bitmap or binary representation of it. I suspect this actually won't work as-is because the deserializer is expecting decoder is expecting it to be a SVG file, not a PNG.

i see that the SVG class itself doesn't have a way to retrieve its source, so we may need to wrap it in our own Svg class that includes the source & the underlying SVG image, then pass that through the pipeline so that we have access to the original SVG source code here.

veena14cs replied:
SvgEncoder is used if Disk caching is necessary in Glide when retrieving data.  If we set DiskCacheStrategy to None we don't need this Encoder and we don't have to manage saving of raw svg file into disk. So I removed this file.

BenHenning replied:
Hmm, I actually assumed we would want Glide to cache images but it may not matter if we're already caching images using the AssetRepository. However, that repository is temporary.

Can you add a TODO comment here on #45 to ensure the caching flow is properly hooked up? I want to make sure that if we check in an implementation that explicitly disables caching that we make sure we're properly caching the images ourselves to avoid continuously redownloading images.

veena14cs replied:
Sure that seems reasonable I will add a TODO in `GlideImageLoader` file to make sure caching is done proper.

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -36,21 +37,40 @@ class UrlImageParser private constructor(
-      gcsPrefix + gcsResource + imageUrl,
-      target
-    )
+    if (imageUrl.endsWith("svg")) {
```
Please also reference the new issue here as a TODO to remove the hardcoded SVG check.

Further, please make this case-agnostic since 'SVG' is also a valid file extension, vs. 'svg'.

veena14cs replied:
Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgSoftwareLayerSetter.kt:
```diff
@@ -0,0 +1,33 @@
+import com.bumptech.glide.request.target.Target
+
+/**
+ * Listener which updates the [ImageView] to be software rendered, because {@link
```
Per my earlier comment thread, I still don't understand why this actually works (e.g. how returning false in both methods below causes it to be software rendered). Do you have additional documentation for this?

veena14cs replied:
Actual implementation of SvgSoftwareLayerSetter is [here](https://gist.github.com/veena14cs/42ad32de4d3b4c0fdb7eb0d6c7c46617). But I am getting class cast exception and also we are not loading in Imageview we are loading in our custom target, so I removed part of the implementation of Imageview which was taking software rendering.

Now I tried to render without adding listener to the glide and svg image is getting loaded, it is making no difference.

BenHenning replied:
Ah, thanks for clarifying. If this ends up being an issue (e.g. we switch one of our layers to use hardware) it looks like we can fix it in our own code by changing the software layer.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/SvgDecoder.kt:
```diff
@@ -0,0 +1,33 @@
+    options: Options
+  ): Resource<SVG?>? {
+    return try {
+      val svg = SVG.getFromInputStream(source)
```
A more Kotlin idiomatic way to do this might be:

```kotlin
SimpleResource(source.use { SVG.getFromInputStream(it) })
```

This automatically closes the input stream, and allows the expression to be cleanly written in a single line without additional variables. Exceptions should be thrown as normal.

veena14cs replied:
Got it. Done.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/RepositoryGlideModule.kt:
```diff
@@ -1,17 +1,23 @@
   override fun registerComponents(context: Context, glide: Glide, registry: Registry) {
-    registry.prepend(ImageAssetFetcher::class.java, InputStream::class.java, RepositoryModelLoader.Factory())
+    // TODO #1039: Introduce custom type OppiaImage for rendering Bitmap and Svg.
+    registry.register(SVG::class.java, PictureDrawable::class.java, SvgDrawableTranscoder())
```
As mentioned earlier (I can't seem to comment on that comment thread anymore), could we pass along a ``Picture`` through the pipeline, instead? Glide doesn't use this until it hits our custom target from what I can tell, so we should be able to pass whatever we want. Since we're just recreating the drawable later anyway, we should probably just pass along the picture object to simplify the pipeline a bit more. Or am I maybe missing something?

veena14cs replied:
We could pass Picture but glide is expecting to have PictureDrawable  for our custom target. So we need to pass drawable.

BenHenning replied:
That makes sense, but aren't we already converting from PictureDrawable to PictureDrawable in the custom target? E.g. in the factory function being passed in for the SVG version of the custom drawable, we're creating a new PictureDrawable using the previous drawable's Picture resource. I'm wondering if we could just pass along the Picture and create the PictureDrawable at the end. Wouldn't that work?

veena14cs replied:
Done. Now it passes Picture instead of PictureDrawable. Thanks for this clarity @BenHenning .

BenHenning reviewed (changes requested):
Thanks @veena14cs! Just a few more comments.

Also, can you make sure all comment threads that are resolved are marked as resolved, otherwise there's a clear question to follow up on them? That makes it easier for me to go back and comment on old threads.

veena14cs committed:
updated new changes

veena14cs committed:
removed encoder file.

veena14cs commented:
> Thanks @veena14cs! Just a few more comments.
> 
> Also, can you make sure all comment threads that are resolved are marked as resolved, otherwise there's a clear question to follow up on them? That makes it easier for me to go back and comment on old threads.

@BenHenning I have addressed your comments also I changed some of the implementation and removed some files which may not be necessary and could load svg's without any issues.

veena14cs committed:
nit changes

veena14cs committed:
Update GlideImageLoader.kt

BenHenning commented on utility/src/main/java/org/oppia/util/parser/GlideImageLoader.kt:
```diff
@@ -27,4 +31,21 @@ class GlideImageLoader @Inject constructor(
+    Glide.with(context)
+      .`as`(PictureDrawable::class.java)
+      .fitCenter()
+      .apply(RequestOptions.diskCacheStrategyOf(DiskCacheStrategy.NONE))
```
Reminder to please add a TODO here for #45 (see my earlier comment).

veena14cs replied:
Done.

BenHenning reviewed (approved):
Thanks @veena14cs! I reopened a couple of the past conversation threads--please resolve those & my new comment before submitting. Otherwise the changes look good. Thanks for implementing this!

If you have any questions about my comments, please ask or reassign the PR to me before submitting.

veena14cs committed:
updated from PictureDrawable to Picture.

veena14cs commented:
> Thanks @veena14cs! I reopened a couple of the past conversation threads--please resolve those & my new comment before submitting. Otherwise the changes look good. Thanks for implementing this!
> 
> If you have any questions about my comments, please ask or reassign the PR to me before submitting.

Changed the implementation as per the discussion. @BenHenning  Please take a final look before I submit this PR.

BenHenning commented on utility/src/main/java/org/oppia/util/parser/UrlImageParser.kt:
```diff
@@ -93,7 +94,7 @@ class UrlImageParser private constructor(
   }
 
-  class UrlDrawable : BitmapDrawable() {
+  class UrlDrawable : BitmapDrawable(){
```
I suggest reverting this (e.g. ``BitmapDrawable() {``).

BenHenning reviewed (approved):
Thanks @veena14cs--this LGTM!

veena14cs committed:
Update UrlImageParser.kt

veena14cs committed:
Update GlideImageLoader.kt
